### PR TITLE
Charm schema

### PR DESCRIPTION
**Pull request by SimonRichardson**
Adds the charm and charm metadata schema in normalized form. This only contains `charm.Meta`, the config, manifest, actions and, lxd-profile will be done in subsequent changes.

Metadata provides, requires and peers are pushed into the same table and are identified via the `charm_relation_kind` table. The deployment field was not brought across as that's being deprecated.

It might be wise at some stage to add a view for just a charm_metadata to get `metadata.yaml` from the normalized form.

Considerations and Omissions:
 - Type: Bundles are not modeled here.
 - Life - Deprecated and removed!
 - Deployment field - Deprecated and removed!
 - Placeholder/pending upload - Placeholder is replaced with the presence of the row. Upload is now replaced with available.
 - Bundlesha256
 - CharmVersion - I'm leaving this out for now, I'm wondering if charm_origin/charm_channel can be used here.
 - StoragePath - The objectstore metadata table will have this information. We could use a link table for this.

### Schema diagram

This shows the schema for the charm tables and associated foreign keys.

model

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

No code should be using this, so the tests should pass. 
We should ensure that we didn't break bootstrap though.

```sh
$ juju bootstrap lxd test
```

## Links

**Jira card:** [JUJU-5852](https://warthogs.atlassian.net/browse/JUJU-5852)

[JUJU-5852]: https://warthogs.atlassian.net/browse/JUJU-5852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/schema/model.go [hunk 1]
```diff
@@ -281,33 +281,395 @@ ON application (name);
...
+        REFERENCES        charm_container(name),
+    PRIMARY KEY (charm_uuid, resource)
+);
+
+CREATE INDEX idx_charm_container_mount_charm
+ON charm_container_mount (charm_uuid);
+
+-- Create a charm url view for backwards compatibility.
+CREATE VIEW v_charm_url AS
+SELECT cs.name || ':' || c.name || '-' || IFNULL(co.revision, 0) AS url, c.uuid AS uuid
```
Comment (SimonRichardson) [inline domain/schema/model.go:669]: I think `SELECT CONCAT(...) AS url...` is more readable, but it turns out CONCAT isn't available everywhere, so we need to fallback to the concat operator (`||`) instead.

File: domain/schema/model/sql/0007-charm.sql [hunk 2]
```diff
@@ -1,27 +1,398 @@
...
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_hash_kind_name
+    ON hash_kind (name);
+
+INSERT INTO hash_kind VALUES
+    (0, 'sha256');
+
+CREATE TABLE charm_hash (
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:127]: I wonder whether we should fold this into the charm_state table (don't necessarily retain that table name). What I mean is that the hash is known when the download is done and the charm is available. So there's a logical grouping of concerns there.

File: domain/schema/model/sql/0007-charm.sql [hunk 3]
```diff
@@ -1,27 +1,398 @@
...
+
+INSERT INTO charm_source VALUES
+    (0, 'local'),
+    (1, 'ch');
+
+CREATE TABLE charm_origin (
+    charm_uuid          TEXT NOT NULL,
+    source_id           INT,
+    id                  string,
+    hash                string,
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:68]: IIANM, we would use this hash to validate the downloaded blob. Do we need to store both this and the hash in the charm_hash table?

File: domain/schema/model/sql/0007-charm.sql [hunk 4]
```diff
@@ -1,27 +1,398 @@
...
+    CONSTRAINT          fk_charm_extra_binding_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_extra_binding_charm
+ON charm_extra_binding (charm_uuid);
+
+CREATE TABLE charm_category (
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:216]: This could do with a comment

File: domain/schema/model/sql/0007-charm.sql [hunk 5]
```diff
@@ -1,27 +1,398 @@
...
+    CONSTRAINT          fk_charm_storage_property_charm_storage
+        FOREIGN KEY     (charm_storage_name)
+        REFERENCES      charm_storage(name),
+    PRIMARY KEY (charm_uuid, charm_storage_name, key)
+);
+
+CREATE INDEX idx_charm_storage_property_charm
+ON charm_storage_property (charm_uuid);
+
+CREATE TABLE charm_device (
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:289]: This was ostensibly for k8s charm and I am not sure if it's used. We should audit what we have on ch and see if we still need this.

File: domain/schema/model/sql/0007-charm.sql [hunk 6]
```diff
@@ -1,27 +1,398 @@
...
+
+CREATE INDEX idx_charm_term_charm
+ON charm_term (charm_uuid);
+
+CREATE TABLE charm_container (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    resource            TEXT,
+    uid                 INT,
+    gid                 INT,
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:365]: These are optional. And we now model them as *int. So we need a way to reflect that in the schema.

File: domain/schema/model/sql/0007-charm.sql [hunk 7]
```diff
@@ -1,27 +1,398 @@
...
+    CONSTRAINT          fk_charm_category_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, value)
 );
 
+CREATE INDEX idx_charm_category_charm
+ON charm_category (charm_uuid);
+
+CREATE TABLE charm_tag (
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:228]: What's the difference between category and tag. "tag" in metadata.yaml is "databases" etc. What's category.

File: domain/schema/model/sql/0007-charm.sql [hunk 8]
```diff
@@ -1,27 +1,398 @@
...
+    resource              TEXT,
+    storage               TEXT,
+    location              TEXT,
+    CONSTRAINT            fk_charm_container_mount_charm
+        FOREIGN KEY       (charm_uuid)
+        REFERENCES        charm(uuid),
+    CONSTRAINT            fk_charm_container_mount_charm_container
+        FOREIGN KEY       (charm_container_name)
+        REFERENCES        charm_container(name),
+    PRIMARY KEY (charm_uuid, resource)
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:387]: PK is charm_uuid, container_name

File: domain/schema/model/sql/0007-charm.sql [hunk 9]
```diff
@@ -1,27 +1,398 @@
...
+);
+
+CREATE INDEX idx_charm_container_charm
+ON charm_container (charm_uuid);
+
+CREATE TABLE charm_container_mount (
+    charm_uuid            TEXT NOT NULL,
+    charm_container_name  TEXT,
+    resource              TEXT,
```
Comment (wallyworld) [inline domain/schema/model/sql/0007-charm.sql:378]: This is part of charm_container

**Review by wallyworld — COMMENTED** [review]
Great start. Some things to discuss.

File: domain/schema/model/sql/0007-charm.sql (hunk 4 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0007-charm.sql:216]: Rename tag

File: domain/schema/model/sql/0007-charm.sql (hunk 3 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0007-charm.sql]: Removed

File: domain/schema/model/sql/0007-charm.sql (hunk 5 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0007-charm.sql:306]: I'm leaving for now after our discussion.

**Review by manadart — APPROVED** [review]
Nice.

**Pull request merged**